File: android/rctmgl/src/main/java/com/mapbox/rctmgl/components/styles/light/RCTMGLLight.java
package com.mapbox.rctmgl.components.styles.light;

import android.content.Context;

import com.facebook.react.bridge.ReadableMap;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.Style;
import com.mapbox.mapboxsdk.style.layers.TransitionOptions;
import com.mapbox.mapboxsdk.style.light.Light;
import com.mapbox.mapboxsdk.style.light.Position;
import com.mapbox.rctmgl.components.AbstractMapFeature;
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView;
import com.mapbox.rctmgl.components.styles.RCTMGLStyle;
import com.mapbox.rctmgl.components.styles.RCTMGLStyleFactory;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by nickitaliano on 9/26/17.
 */

public class RCTMGLLight extends AbstractMapFeature {
    private MapboxMap mMap;
    private ReadableMap mReactStyle;

    public RCTMGLLight(Context context) {
        super(context);
    }

    @Override
    public void addToMap(RCTMGLMapView mapView) {
        mMap = mapView.getMapboxMap();
        setLight();
    }

    @Override
    public void removeFromMap(RCTMGLMapView mapView) {
        // ignore there's nothing to remove just update the light style
    }

    public void setReactStyle(ReadableMap reactStyle) {
        mReactStyle = reactStyle;

        setLight();
    }

    private void setLight(Light light) {
        RCTMGLStyleFactory.setLightLayerStyle(light, new RCTMGLStyle(getContext(), mReactStyle, mMap));
    }

    private void setLight() {
        Style style = getStyle();
        if (style != null) {
            setLight(style.getLight());
        }
    }

    private Style getStyle() {
        if (mMap == null) {
            return null;
        }
        return mMap.getStyle();
    }
}

File: android/src/main/java-v10/com/mapbox/rctmgl/components/styles/light/RCTMGLLight.kt
package com.mapbox.rctmgl.components.styles.light

import android.content.Context
import com.mapbox.rctmgl.components.AbstractMapFeature
import com.mapbox.maps.MapboxMap
import com.facebook.react.bridge.ReadableMap
import com.mapbox.maps.Style
import com.mapbox.rctmgl.components.RemovalReason
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView
import com.mapbox.rctmgl.components.styles.RCTMGLStyleFactory
import com.mapbox.rctmgl.components.styles.RCTMGLStyle

import com.mapbox.rctmgl.v11compat.light.*

class RCTMGLLight(context: Context?) : AbstractMapFeature(context) {
    private var mMap: MapboxMap? = null
    private var mReactStyle: ReadableMap? = null
    override fun addToMap(mapView: RCTMGLMapView) {
        super.addToMap(mapView)
        mMap = mapView.getMapboxMap()
        setLight()
    }

    override fun removeFromMap(mapView: RCTMGLMapView, reason: RemovalReason): Boolean {
        // ignore there's nothing to remove just update the light style
        return super.removeFromMap(mapView, reason)
    }

    fun setReactStyle(reactStyle: ReadableMap?) {
        mReactStyle = reactStyle
        setLight()
    }

    private fun setLight(light: Light) {
        RCTMGLStyleFactory.setLightLayerStyle(light, RCTMGLStyle(context, mReactStyle!!, mMap!!))
        style!!.setLight(light)
    }

    private fun setLight() {
        if (style != null) {
            val light = createLight()
            setLight(light)
        }
    }

    private val style: Style?
        private get() = if (mMap == null) {
            null
        } else mMap!!.getStyle()
}